src/program_state.cpp

```diff
@@ -513,15 +513,22 @@ hsa_executable_t load_executable(const string& file, hsa_executable_t executable
     return executable;
 }
 
-// To force HIP to load the kernels and to setup the function
-// symbol map on program startup
+// HIP startup kernel loader logic
+// When enabled HIP_STARTUP_LOADER, HIP will load the kernels and setup
+// the function symbol map on program startup
 class startup_kernel_loader {
    private:
     startup_kernel_loader() { functions(); }
     startup_kernel_loader(const startup_kernel_loader&) = delete;
     startup_kernel_loader& operator=(const startup_kernel_loader&) = delete;
     static startup_kernel_loader skl;
```

Comment:
where would `    static startup_kernel_loader skl;` be instantiated? if it's not instantiated anywhere should this be removed?

Reply:
another idea is to let `skl` be a pointer, like: `static startup_kernel_loader*`.

and in the shared lib constructor, instead of calling `functions()` directly, use `new skl` to make sure the default constructor of `startup_kernel_loader()` is invoked, and `*skl` points to the instance of the startup kernel loader.